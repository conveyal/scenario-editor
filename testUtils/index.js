export function mockComponents (components) {
  if (components instanceof Array) {
    const mockedComponents = {}
    for (let i = 0; i < components.length; i++) {
      const name = components[i]
      mockedComponents[name] = name
    }
    return mockedComponents
  }
}
