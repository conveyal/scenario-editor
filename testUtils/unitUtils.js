/* globals describe, expect, it */

import React from 'react'
import renderer from 'react-test-renderer'

/**
 * Performs a basic render test of a Component
 *
 * @param  {Mixed} components Can be either an array of objects or a single object
 *   Each object can have the following properties:
 *   - children: Any children to render within the component
 *   - component (required): The component to render
 *   - name: How to classify the test in the `describe section`
 *   - notToBeCalledFns: An array of String keys that map to property values that are jest.fn()'s that will be asserted as not called during basic render test
 *   - props (required): The props of the component to render
 *   - toBeCalledFns: Same as `notToBeCalledFns` except they are asserted to be called
 */
export function basicRenderTest (components) {
  // prepare components to be tested
  if (!Array.isArray(components)) {
    components = [components]
  }

  // run tests on each component
  components.forEach((component) => {
    const Component = component.component

    if (typeof component.name !== 'string') {
      throw new Error('Component name not provided in basicRenderTest cfg')
    }

    describe(`Component > ${component.name}`, () => {
      it('renders correctly', () => {
        const tree = renderer.create(
          <Component
            {...component.props}
            >
            {component.children}
          </Component>
        ).toJSON()
        expect(tree).toMatchSnapshot()

        if (Array.isArray(component.toBeCalledFns)) {
          component.toBeCalledFns.forEach((fn) => {
            expect(component.props[fn]).toBeCalled()
          })
        }

        if (Array.isArray(component.notToBeCalledFns)) {
          component.notToBeCalledFns.forEach((fn) => {
            expect(component.props[fn]).not.toBeCalled()
          })
        }
      })
    })
  })
}

/**
 * Creates a object to mock a module with multiple exports
 *
 * @param  {Array} components An array of Strings representing the named exports
 * @return {Object}           A mock representation of the named exports
 */
export function mockExports (components) {
  if (Array.isArray(components)) {
    const mockedComponents = {}
    for (let i = 0; i < components.length; i++) {
      const name = components[i]
      mockedComponents[name] = name
    }
    return mockedComponents
  }
}
