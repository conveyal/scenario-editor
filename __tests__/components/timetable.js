/* global describe, it, expect, jest */

import React from 'react'
import renderer from 'react-test-renderer'

import { mockComponents } from '../../testUtils'

jest.mock('../../lib/components/buttons', () => { return mockComponents(['Button']) })
jest.mock('../../lib/components/icon', () => 'Icon')
jest.mock('../../lib/components/input', () => { return mockComponents(['Number', 'Text']) })
jest.mock('../../lib/components/timetable-entry', () => 'TimetableEntry')

import Timetable from '../../lib/components/timetable'

describe('Timetable', () => {
  it('renders correctly', () => {
    const timetable = {
      name: 'Test timetable',
      speed: 40,
      dwellTime: 10,
      monday: true,
      tuesday: true,
      wednesday: true,
      thursday: true,
      friday: true,
      saturday: false,
      sunday: false,
      headwaySecs: 900, // 15 minutes
      startTime: 28800, // 8am
      endTime: 57600 // 4pm
    }
    const removeTimetableFn = jest.fn()
    const replaceTimetableFn = jest.fn()
    const tree = renderer.create(
      <Timetable
        timetable={timetable}
        removeTimetable={removeTimetableFn}
        replaceTimetable={replaceTimetableFn}
        />
    ).toJSON()
    expect(tree).toMatchSnapshot()
    expect(removeTimetableFn).not.toBeCalled
    expect(replaceTimetableFn).not.toBeCalled
  })
})
