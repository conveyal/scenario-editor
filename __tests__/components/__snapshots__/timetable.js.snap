exports[`Timetable renders correctly 1`] = `
<section
  className="panel panel-default inner-panel">
  <div
    className="panel-heading clearfix">
    <strong>
      Test timetable
    </strong>
    <Button
      className="pull-right"
      onClick={[Function mockConstructor]}
      size="sm"
      style="danger"
      title="Delete timetable">
      <Icon
        type="close" />
    </Button>
  </div>
  <div
    className="panel-body">
    <Text
      name="Name"
      onChange={[Function anonymous]}
      value="Test timetable" />
    <TimetableEntry
      replaceTimetable={[Function mockConstructor]}
      timetable={
        Object {
          "dwellTime": 10,
          "endTime": 57600,
          "friday": true,
          "headwaySecs": 900,
          "monday": true,
          "name": "Test timetable",
          "saturday": false,
          "speed": 40,
          "startTime": 28800,
          "sunday": false,
          "thursday": true,
          "tuesday": true,
          "wednesday": true
        }
      } />
    <Number
      label="Speed (km/h)"
      onChange={[Function anonymous]}
      value={40} />
    <Number
      label="Dwell time (seconds)"
      onChange={[Function anonymous]}
      value={10} />
  </div>
</section>
`;
